internal/olm/operator/registry/olm_resources.go

```diff
@@ -68,6 +68,7 @@ func newSubscription(csvName, namespace string, opts ...func(*v1alpha1.Subscript
 	sub.SetName(getSubscriptionName(csvName))
 	sub.SetNamespace(namespace)
 	sub.Spec = &v1alpha1.SubscriptionSpec{}
+	// TODO(asmacdo) This pattern freaks me out. Ask someone about it.
 	for _, opt := range opts {
 		opt(sub)
 	}
```

Comment:
It is quite weird. Basically it's a functional option pattern. https://www.sohamkamani.com/golang/options-pattern/ The idea is that you can supply optional items when constructing an instance. Often these options are of very different types, as in the example of the `House` in the url above they wanted to pass in `WithConcrete` which is a `string` and `WithFireplace` which is a `bool`.

WIthout this pattern, you would either have to have a lengthy constructor function that you always have to pass in every option (even if they are optional). i.e. `NewHouse(material string, hasfireplace bool, floors int, foo Foo, bar *Bar, etc)`. It also means every time you want to extend it you'd likely have to add more to the constructor function.

With the options pattern the option is a function. In our case for subscription we did not define a new Option type. If we did it would look like this: `type SubscriptionOption func(*v1alpha1.Subscription)`. So we're just returning a function that takes in a `*v1alpha1.Subscription`. 

So with this pattern, we can create `WithSomeOption` functions that take in the specific type we want and by returning a function it can be easily used by the constructor function. So in the above code you see we have the following:

```golang
func withInstallPlanApproval(approval v1alpha1.Approval) func(*v1alpha1.Subscription) {
	return func(sub *v1alpha1.Subscription) {
		if sub.Spec == nil {
			sub.Spec = &v1alpha1.SubscriptionSpec{}
		}
		sub.Spec.InstallPlanApproval = approval
	}
}
```
That will take in an Approval which is an attribute we want to set on the `Subscription`. So we return a function that expects a `Subscription` to be passed in so we can modify it. If I want to override the `PackageName` I could easily create the following to update `PackageName`:

```golang
func withPackageName(pkgName string) func(*v1alpha1.Subscription) {
	return func(sub *v1alpha1.Subscription) {
		if sub.Spec == nil {
			sub.Spec = &v1alpha1.SubscriptionSpec{}
		}
		sub.Spec.PackageName = pkgName
	}
}
```

I don't have to change the constructor function since it takes in a variable list of "options" which is a list of `func(*v1alpha1.Subscription)`.

Reply:
Thats pretty slick! Thanks!